Issue2653 stop nominating once confirmed

hewison-chris opened the pull request:
If the node waits till it has gossiped an envelope to all connected peers then it may well end up nominating new data sets when it should not.

hewison-chris committed:
Many transactions unit test

This is a network unit test which keeps increasing the number of
transactions sent to be externalized. This is to test the gossiping and
consensus protocol implementation works with large numbers of
transactions.

hewison-chris committed:
Add ansi colors for verbose and debug logging

hewison-chris committed:
Set confirmed height earlier

If we set the confirmed height after gossipping our envelope then the
nomination timer could still fire whilst we are gossipping. This would
result in another nomination being made. If we set the confirmed height
before gossipping this is prevented.

hewison-chris committed:
Debug logging for `attemptRequest`

hewison-chris committed:
Remove signConfirmBallot function

This function was created when we were signing the block header also at
the time we signed a confirm ballot. This is no longer the case and the
function does not serve any purpose.

hewison-chris committed:
Reduce the frequency of refreshing connections

By using the `network_discovery_interval` duration which is usually set
to a larger duration than the `retry_delay`.

hewison-chris committed:
Refactor the wait for more than half signed last block

Previously the check was done on receiving a `nominate` envelope and
then skipped processing this envelope. Now it is checked when a
`confirm` envelope is received and waits until there are enough
signatures added before processing this `confirm` envelope.

hewison-chris committed:
Use the `preimage_catchup_interval` for catchup timer

The timer was incorrectly armed using the `preimage_reveal_interval`.

hewison-chris committed:
Reduce interval for preimage catchup in unit tests

omerfirmak commented on source/agora/network/Client.d:
```diff
@@ -570,9 +575,15 @@ public class NetworkClient
                 this.taskman.wait(this.retry_delay);
+            }
         }
+        log.info("Client.attemptRequest '{}' to all {} peers FAILED after {} retires", name, this.connections.length, this.max_retries);
```
retires

omerfirmak commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -646,27 +646,38 @@ extern(D):
+                    last_block.header.validators.setCount);
+                long waited = 0;
+                // Wait till more than half have signed previous block
+                while (1)
```
Instead of blocking and waiting for the signatures here, we probably should put it back to the `queued_envelopes` queue and let the `envelopeProcessTask` handle it again.

hewison-chris replied:
I was thinking about that too. We should probably also do the same for the missing `tx` case so that we do not throw away that envelope. It would also be nice if we could trigger the tasks here to fetch signatures (or missing `tx`s) so we don;t have to wait for the timer to fire.

hewison-chris committed:
Do not throw away commit envelopes when missing tx

If we recieve an envelope with tx set containing any tx we do not yet
have in the pool then retry this envelope again later.

hewison-chris commented:
Closing this PR as it is now split into many.
